src/sources/exec.rs

```diff
@@ -0,0 +1,886 @@
+    internal_events::{ExecEventReceived, ExecFailed},
+    shutdown::ShutdownSignal,
+    Pipeline,
+};
+use bytes::{Bytes, BytesMut};
+use chrono::Utc;
+use futures::{FutureExt, SinkExt, StreamExt};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+use std::io;
+use std::io::{Error, ErrorKind};
+use std::path::PathBuf;
+use std::process::ExitStatus;
+use tokio::io::{AsyncRead, BufReader};
+use tokio::process::Command;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::time::{self, sleep, Duration, Instant};
+use tokio_stream::wrappers::IntervalStream;
+use tokio_util::codec::{Decoder, FramedRead, LinesCodec};
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default, deny_unknown_fields)]
+pub struct ExecConfig {
+    pub mode: Mode,
+    pub scheduled: Option<ScheduledConfig>,
+    pub streaming: Option<StreamingConfig>,
+    pub command: Vec<String>,
+    pub working_directory: Option<PathBuf>,
+    #[serde(default = "default_include_stderr")]
+    pub include_stderr: bool,
+    #[serde(default = "default_events_per_line")]
+    pub event_per_line: bool,
+    #[serde(default = "default_maximum_buffer_size")]
+    pub maximum_buffer_size_bytes: usize,
+}
+
+// TODO: Would be nice to combine the scheduled and streaming config with the mode enum once
+//       this serde ticket has been addressed (https://github.com/serde-rs/serde/issues/2013)
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub enum Mode {
+    Scheduled,
+    Streaming,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct ScheduledConfig {
+    #[serde(default = "default_exec_interval_secs")]
+    exec_interval_secs: u64,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct StreamingConfig {
+    #[serde(default = "default_respawn_on_exit")]
+    respawn_on_exit: bool,
+    #[serde(default = "default_respawn_interval_secs")]
+    respawn_interval_secs: u64,
+}
+
+#[derive(Debug, PartialEq, Snafu)]
+pub enum ExecConfigError {
+    #[snafu(display("A non-empty list for command must be provided"))]
+    CommandEmpty,
+    #[snafu(display("The maximum buffer size must be greater than zero"))]
+    ZeroBuffer,
+}
+
+impl Default for ExecConfig {
+    fn default() -> Self {
+        ExecConfig {
+            mode: Mode::Scheduled,
+            scheduled: Some(ScheduledConfig {
+                exec_interval_secs: default_exec_interval_secs(),
+            }),
+            streaming: None,
+            command: vec!["echo".to_owned(), "Hello World!".to_owned()],
+            working_directory: None,
+            include_stderr: default_include_stderr(),
+            event_per_line: default_events_per_line(),
+            maximum_buffer_size_bytes: default_maximum_buffer_size(),
+        }
+    }
+}
+
+// TODO: Maybe split into a shared codec module
```

Comment:
If you felt like it, you could move it into a separate file in `src/sources/exec/` and rename this one `src/sources/exec/mod.rs`. We could extract it later if we find a use for it outside of the `exec` source.

I think it would be nice to have some tests specifically for the codec.

Reply:
Created the separate file with tests